internal/rpcapi/stacks_test.go

```diff
@@ -969,6 +976,287 @@ func TestStackChangeProgress(t *testing.T) {
+		return stacks_testing_provider.NewProvider(t), nil
+	}
+
+	lock := depsfile.NewLocks()
+	lock.SetProvider(
+		addrs.NewDefaultProvider("testing"),
+		providerreqs.MustParseVersion("0.0.0"),
+		providerreqs.MustParseVersionConstraints("=0.0.0"),
+		providerreqs.PreferredHashes([]providerreqs.Hash{}),
+	)
+	lockHandle := handles.NewDependencyLocks(lock)
+
+	stream, err := stacksClient.MigrateTerraformState(ctx, &stacks.MigrateTerraformState_Request{
+		StateHandle:           resp.StateHandle,
+		ConfigHandle:          openResp.StackConfigHandle,
+		DependencyLocksHandle: lockHandle.ForProtobuf(),
+		Mapping: &stacks.MigrateTerraformState_Request_Simple{
+			Simple: &stacks.MigrateTerraformState_Request_Mapping{
+				ResourceAddressMap: map[string]string{
+					"testing_deferred_resource.resource": "self",
+				},
+			},
+		},
+	})
+
+	if err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+	gotEvents := []*stacks.MigrateTerraformState_Event{}
+	for {
+		event, err := stream.Recv()
+		if err == io.EOF {
+			break
+		}
+
+		if err != nil {
+			t.Fatalf("unexpected error: %s", err)
+		}
+
+		gotEvents = append(gotEvents, event)
+	}
+
+	wantChanges := []*stacks.AppliedChange_ChangeDescription{
+		{
+			Description: &stacks.AppliedChange_ChangeDescription_ResourceInstance{
+				ResourceInstance: &stacks.AppliedChange_ResourceInstance{
+					Addr: &stacks.ResourceInstanceObjectInStackAddr{
+						ComponentInstanceAddr: "component.self",
+						ResourceInstanceAddr:  "testing_deferred_resource.resource",
+					},
+					ResourceMode: stacks.ResourceMode_MANAGED,
+					ResourceType: "testing_deferred_resource",
+					ProviderAddr: "registry.terraform.io/hashicorp/testing",
+				},
+			},
+		},
+		{
+			Description: &stacks.AppliedChange_ChangeDescription_ComponentInstance{
+				ComponentInstance: &stacks.AppliedChange_ComponentInstance{
+					ComponentAddr:         "component.self",
+					ComponentInstanceAddr: "component.self",
+				},
+			},
+		},
+	}
+
+	if len(gotEvents) != len(wantChanges) {
+		t.Fatalf("expected %d events, got %d", len(wantChanges), len(gotEvents))
+	}
+
+	gotChanges := make([]*stacks.AppliedChange_ChangeDescription, len(gotEvents))
+	for i, evt := range gotEvents {
+		gotChanges[i] = evt.GetAppliedChange().Descriptions[0]
+	}
+
+	_cmpopts := []cmp.Option{cmpopts.IgnoreUnexported(
+		stacks.AppliedChange_ChangeDescription{},
+		stacks.AppliedChange_ResourceInstance{},
+		stacks.AppliedChange_ComponentInstance{},
+		stacks.AppliedChange_ChangeDescription_ResourceInstance{},
+		stacks.AppliedChange_ChangeDescription_ComponentInstance{},
+		stacks.ResourceInstanceObjectInStackAddr{},
+	),
+		cmpopts.IgnoreFields(stacks.AppliedChange_ChangeDescription{}, "Key"),
+		cmpopts.IgnoreFields(stacks.AppliedChange_ResourceInstance{}, "NewValue"),
+		cmpopts.IgnoreTypes(&stacks.DynamicValue{}, protoimpl.MessageState{}, protoimpl.UnknownFields{}),
+	}
```

Comment:
Do you think using `protocmp` could tidy this up a bit? https://pkg.go.dev/google.golang.org/protobuf/testing/protocmp